zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs

```diff
@@ -0,0 +1,522 @@
+            callee_address.clone(),
+            is_warm.expr(),
+            is_warm_prev.expr(),
+            Some(&mut reversion_info),
+        );
+
+        // Verify transfer
+        let value_is_zero = IsZeroGadget::construct(cb, sum::expr(&value.cells));
+        let has_value = 1.expr() - value_is_zero.expr();
+        cb.condition(has_value.clone(), |cb| {
+            cb.require_zero(
+                "CALL with value must not be in static call stack",
+                is_static.expr(),
+            );
+        });
+
+        cb.account_read(
+            callee_address.clone(),
+            AccountFieldTag::Balance,
+            balance.expr(),
+        );
+
+        // Verify gas cost
+        let [callee_nonce, callee_code_hash] = [AccountFieldTag::Nonce, AccountFieldTag::CodeHash]
+            .map(|field_tag| {
+                let value = cb.query_cell();
+                cb.account_read(callee_address.clone(), field_tag, value.expr());
+                value
+            });
+        let is_empty_nonce_and_balance =
+            BatchedIsZeroGadget::construct(cb, [callee_nonce.expr(), balance.expr()]);
+        let is_empty_code_hash = IsEqualGadget::construct(
+            cb,
+            callee_code_hash.expr(),
+            Word::random_linear_combine_expr(
+                (*EMPTY_HASH_LE).map(|byte| byte.expr()),
+                cb.power_of_randomness(),
+            ),
+        );
+        let is_empty_account = is_empty_nonce_and_balance.expr() * is_empty_code_hash.expr();
+        // Sum up gas cost
+        let gas_cost = select::expr(
+            is_warm_prev.expr(),
+            GasCost::WARM_ACCESS.expr(),
+            GasCost::COLD_ACCOUNT_ACCESS.expr(),
+        ) + has_value
+            * (GasCost::CALL_WITH_VALUE.expr() + is_empty_account * GasCost::NEW_ACCOUNT.expr())
+            + memory_expansion.gas_cost();
+
+        // Check if the amount of gas available is less than the amount of gas
+        // required
+        let insufficient_gas = LtGadget::construct(cb, cb.curr.state.gas_left.expr(), gas_cost);
+        cb.require_equal(
+            "constant gas left is less than gas required ",
+            insufficient_gas.expr(),
+            1.expr(),
+        );
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+
+        // Go to EndTx only when is_root
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
+        );
+
+        // When it's a root call
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(18.expr() + cb.curr.state.reversible_write_counter.expr()),
```

Comment:
This part is different from the spec. https://github.com/privacy-scaling-explorations/zkevm-specs/blob/0bfb00a41cc45564fe1c111c926cb8b028cfb053/src/zkevm_specs/evm/execution/oog_call.py#L94

Reply:
it is related to discussion https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/827#discussion_r1031341590 , will update after it finalizes.

Reply:
will update spec to match the real circuit .

Reply:
spec PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/332 will update to match circuit diff.